### PR TITLE
Add function pointers for every type of rectangle

### PR DESCRIPTION
**Pull request by ldmnyblzs**
These changes make it possible for the application to handle all of the rendering and JPEG decoding, thus allowing GPU accelerated rendering and multi-core decoding. I'm using this on a Raspberry Pi 2 where the CPU is not powerful enough to handle VNC.

### CODE REVIEW

File: rfb/rfbclient.h [hunk 1]
```diff
@@ -307,6 +310,9 @@ typedef struct _rfbClient {
 
 	GotCursorShapeProc GotCursorShape;
 	GotCopyRectProc GotCopyRect;
+  GotFillRectProc GotFillRect;
```
Comment (bk138) [inline rfb/rfbclient.h]: Please add those at the end of the struct to not break ABI compatibility.

File: rfb/rfbclient.h [hunk 2]
```diff
@@ -307,6 +310,9 @@ typedef struct _rfbClient {
...
+  GotBitmapProc GotBitmap;
```
Comment (bk138) [inline rfb/rfbclient.h]: Also, can you please add some inline documention what those hooks do?

File: libvncclient/corre.c [hunk 3]
```diff
@@ -46,7 +46,10 @@ HandleCoRREBPP (rfbClient* client, int rx, int ry, int rw, int rh)
     if (!ReadFromRFBServer(client, (char *)&pix, sizeof(pix)))
 	return FALSE;
 
-    FillRectangle(client, rx, ry, rw, rh, pix);
+    if(client->GotFillRect != NULL)
```
Comment (bk138) [inline libvncclient/corre.c]: Just a thought: You could get rid of those ifs by connecting the original functions to the hooks in vncviewer.c:rfbGetClient()

**Comment by bk138** [comment]
Thanks! I added some comments.

File: rfb/rfbclient.h [hunk 4]
```diff
@@ -373,6 +370,18 @@ typedef struct _rfbClient {
 	LockWriteToTLSProc LockWriteToTLS;
 	UnlockWriteToTLSProc UnlockWriteToTLS;
 
+        /* Hooks for custom rendering
```
Comment (bk138) [inline rfb/rfbclient.h]: Sry, I should've mentioned that we use doxygen --> https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html So you have to start comments with /**

**Comment by ldmnyblzs** [comment]
I moved the hooks to the end of the struct and added some comments.
I tried to implement the hooks as similarly to the already existing GotCopyRect hook as possible. That one is invoked in an if-else statement. Should I connect all of these hooks (including GotCopyRect) to the original functions in vncviewer.c:rfbGetClient() instead?

**Comment by bk138** [comment]
@dscho what do you think?

File: libvncclient/corre.c (hunk 3 above)
Comment (dscho) [inline libvncclient/corre.c]: I agree. It would be much nicer to initialize the `GotFillRect` attribute to `FillRectangle` in `rfbGetClient()`.

File: libvncclient/ultra.c [hunk 5]
```diff
@@ -199,8 +202,11 @@ HandleUltraZipBPP (rfbClient* client, int rx, int ry, int rw, int rh)
 
     if (se == rfbEncodingRaw)
     {
+      if (client->GotBitmap != NULL)
+	client->GotBitmap(client, (unsigned char *)ptr, sx, sy, sw, sh);
+      else
         CopyRectangle(client, (unsigned char *)ptr, sx, sy, sw, sh);
-        ptr += ((sw * sh) * (BPP / 8));
+      ptr += ((sw * sh) * (BPP / 8));
     }
```
Comment (dscho) [inline libvncclient/ultra.c]: This is a separate change and needs to be in its own commit.

File: rfb/rfbclient.h [hunk 6]
```diff
@@ -373,6 +370,18 @@ typedef struct _rfbClient {
...
+         * VNC rendering boils down to 3 activities:
+         * - GotCopyRect: copy an area of the framebuffer
+         * - GotFillRect: fill an area of the framebuffer with a solid color
+         * - GotBitmap: copy the bitmap in the buffer into the framebuffer
+	 * The client application should either set all three of these or none!
+         */
+        GotFillRectProc GotFillRect;
+        GotBitmapProc GotBitmap;
+        /** Hook for custom JPEG decoding and rendering */
+        GotJpegProc GotJpeg;
```
Comment (dscho) [inline rfb/rfbclient.h]: It would be better to amend the original commit, rather than adding the fields in the wrong place only to fix that problem in the very same PR. Let's not introduce problems unless necessary :grinning:

File: rfb/rfbclient.h [hunk 7]
```diff
@@ -370,13 +370,13 @@ typedef struct _rfbClient {
...
 
-        /* Hooks for custom rendering
+        /** Hooks for custom rendering
          *
          * VNC rendering boils down to 3 activities:
          * - GotCopyRect: copy an area of the framebuffer
          * - GotFillRect: fill an area of the framebuffer with a solid color
          * - GotBitmap: copy the bitmap in the buffer into the framebuffer
-	 * The client application should either set all three of these or none!
+         * The client application should either set all three of these or none!
```
Comment (dscho) [inline rfb/rfbclient.h]: Again, this is a problem that was introduced by the same PR that fixes it. That is not necessary. We can all pretend that no problem was introduced, ever, and that we are geniuses who never introduce problems but only fix them.

In other words: why not squash this change, as well as the previous one, into the first commit?

**Comment by bk138** [comment]
@ldmnyblzs OK, so please connect the hooks  to the original functions in vncviewer.c:rfbGetClient() Thanks!

File: rfb/rfbclient.h (hunk 7 above)
Comment (ldmnyblzs) [inline rfb/rfbclient.h]: Can you give me more detailed instructions on how to do that? This is the first time I try to contribute to a project, so I'm not familiar with the procedure.

**Comment by ldmnyblzs** [comment]
Oops! The original functions are defined in libvncclient/rfbproto.c. Their declarations should be put into one of the headers, but which one?

**Comment by dscho** [comment]
> The original functions are defined in libvncclient/rfbproto.c. Their declarations should be put into one of the headers, but which one?

Since they are declared `static` and are not used in `rfbproto.c` (or in any of the files included from there, such as `hextile.c`) anymore, the definitions should move into `libvncserver.c`, because that is now the sole user.

File: libvncclient/rre.c [hunk 8]
```diff
@@ -59,7 +59,8 @@ HandleRREBPP (rfbClient* client, int rx, int ry, int rw, int rh)
     subrect.w = rfbClientSwap16IfLE(subrect.w);
     subrect.h = rfbClientSwap16IfLE(subrect.h);
 
-    FillRectangle(client, rx+subrect.x, ry+subrect.y, subrect.w, subrect.h, pix);
+    client->GotFillRect(client, rx+subrect.x, ry+subrect.y, subrect.w,
```
Comment (bk138) [inline libvncclient/rre.c]: Really picky minor thing: pls don't add a line break here.

File: libvncclient/tight.c [hunk 9]
```diff
@@ -277,7 +277,8 @@ HandleTightBPP (rfbClient* client, int rx, int ry, int rw, int rh)
       if (extraBytes > 0)
 	memcpy(client->buffer, &client->buffer[numRows * rowSize], extraBytes);
 
-      CopyRectangle(client, (uint8_t *)buffer2, rx, ry+rowsProcessed, rw, numRows);
+      client->GotBitmap(client, (uint8_t *)buffer2, rx, ry+rowsProcessed, rw,
```
Comment (bk138) [inline libvncclient/tight.c]: Same here: pls no line break.
